drivers/stmpe811/Kconfig

```diff
@@ -9,32 +9,32 @@ config MODULE_STMPE811
     bool
     depends on HAS_PERIPH_GPIO
     depends on HAS_PERIPH_GPIO_IRQ
-    depends on HAS_PERIPH_I2C
     depends on TEST_KCONFIG
     select MODULE_PERIPH_GPIO
     select MODULE_PERIPH_GPIO_IRQ
     select MODULE_ZTIMER
     select MODULE_ZTIMER_MSEC
-    depends on TEST_KCONFIG
-
-choice
-    bool "STMPE811 Touchscreen Controller"
-    optional
-    depends on TEST_KCONFIG
-    help
-        The driver supports both STMPE811 connected either via SPI or
-        I2C bus. Select one combination.
 
 config MODULE_STMPE811_I2C
-    bool "STMPE811 on I2C"
+    bool "STMPE811 Touchscreen Controller on I2C"
     depends on HAS_PERIPH_I2C
     select MODULE_PERIPH_I2C
     select MODULE_STMPE811
 
 config MODULE_STMPE811_SPI
-    bool "STMPE811 on SPI"
+    bool "STMPE811 Touchscreen Controller on SPI"
     depends on HAS_PERIPH_SPI
     select MODULE_PERIPH_SPI
     select MODULE_STMPE811
 
-endchoice
+config HAVE_STMPE811_I2C
```

Comment:
I think we are deciding on some newer design patterns for describing hardware connected to the board, this is similar to `drivers/bmx280/Kconfig` or `drivers/isl29020/Kconfig`.  The idea is to try to keep the logic output the `HAVE_*` and keep it in the module, selecting the module with `default (MODULE_TOUCH_DEV && HAVE_STMPE811)`.

Here is a diff of what I would think...

```diff
diff --git a/drivers/stmpe811/Kconfig b/drivers/stmpe811/Kconfig
index 4fdf511400..7fe4757a94 100644
--- a/drivers/stmpe811/Kconfig
+++ b/drivers/stmpe811/Kconfig
@@ -5,8 +5,10 @@
 # directory for more details.
 #
 
-config MODULE_STMPE811
+menuconfig MODULE_STMPE811
     bool
+    prompt "STMPE811 Touchscreen Controller" if !(MODULE_TOUCH_DEV && HAVE_STMPE811)
+    default (MODULE_TOUCH_DEV && HAVE_STMPE811)
     depends on HAS_PERIPH_GPIO
     depends on HAS_PERIPH_GPIO_IRQ
     depends on TEST_KCONFIG
@@ -15,26 +17,45 @@ config MODULE_STMPE811
     select MODULE_ZTIMER
     select MODULE_ZTIMER_MSEC
 
+if MODULE_STMPE811
+
+choice STMPE811_VARIANT
+    bool "Model"
+    default MODULE_STMPE811_I2C if HAVE_STMPE811_I2C
+    default MODULE_STMPE811_SPI if HAVE_STMPE811_SPI
+    help
+        The driver supports both stmpe811 connected either via SPI or
+        I2C bus. Select one combination.
+
 config MODULE_STMPE811_I2C
-    bool "STMPE811 Touchscreen Controller on I2C"
+    bool "STMPE811 on I2C"
     depends on HAS_PERIPH_I2C
     select MODULE_PERIPH_I2C
-    select MODULE_STMPE811
 
 config MODULE_STMPE811_SPI
-    bool "STMPE811 Touchscreen Controller on SPI"
+    bool "STMPE811 on SPI"
     depends on HAS_PERIPH_SPI
+    depends on HAS_PERIPH_GPIO
     select MODULE_PERIPH_SPI
-    select MODULE_STMPE811
+    select MODULE_PERIPH_GPIO
+
+endchoice
+
+endif # MODULE_STMPE811
+
+config HAVE_STMPE811
+    bool
+    help
+      Indicates that an STMPE811 is present.
 
 config HAVE_STMPE811_I2C
     bool
-    select MODULE_STMPE811_I2C if MODULE_TOUCH_DEV
+    select HAVE_STMPE811
     help
       Indicates that an STMPE811 over I2C touch panel is present.
 
 config HAVE_STMPE811_SPI
     bool
-    select MODULE_STMPE811_SPI if MODULE_TOUCH_DEV
+    select HAVE_STMPE811
     help
       Indicates that an STMPE811 over SPI touch panel is present.
```

Reply:
Thanks! I applied your suggestion (directly squashed) and it's still working.

<details>

```
$ TEST_KCONFIG=1 make -C tests/touch_dev/ flash term
make: Entering directory '/work/riot/RIOT/tests/touch_dev'
=== [ATTENTION] Testing Kconfig dependency modelling ===
Building application "tests_touch_dev" for "stm32f429i-disc1" with MCU "stm32".

"make" -C /work/riot/RIOT/boards/stm32f429i-disc1
"make" -C /work/riot/RIOT/core
"make" -C /work/riot/RIOT/cpu/stm32
"make" -C /work/riot/RIOT/cpu/cortexm_common
"make" -C /work/riot/RIOT/cpu/cortexm_common/periph
"make" -C /work/riot/RIOT/cpu/stm32/periph
"make" -C /work/riot/RIOT/cpu/stm32/stmclk
"make" -C /work/riot/RIOT/cpu/stm32/vectors
"make" -C /work/riot/RIOT/drivers
"make" -C /work/riot/RIOT/drivers/periph_common
"make" -C /work/riot/RIOT/drivers/stmpe811
"make" -C /work/riot/RIOT/drivers/touch_dev
"make" -C /work/riot/RIOT/sys
"make" -C /work/riot/RIOT/sys/auto_init
"make" -C /work/riot/RIOT/sys/auto_init/screen
"make" -C /work/riot/RIOT/sys/frac
"make" -C /work/riot/RIOT/sys/malloc_thread_safe
"make" -C /work/riot/RIOT/sys/newlib_syscalls_default
"make" -C /work/riot/RIOT/sys/pm_layered
"make" -C /work/riot/RIOT/sys/stdio_uart
"make" -C /work/riot/RIOT/sys/ztimer
   text	   data	    bss	    dec	    hex	filename
  14496	    156	   2556	  17208	   4338	/work/riot/RIOT/tests/touch_dev/bin/stm32f429i-disc1/tests_touch_dev.elf
/work/riot/RIOT/dist/tools/openocd/openocd.sh flash /work/riot/RIOT/tests/touch_dev/bin/stm32f429i-disc1/tests_touch_dev.elf
### Flashing Target ###
Open On-Chip Debugger 0.11.0+dev-00551-gaad871805 (2022-01-06-14:44)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
hla_swd
Info : The selected transport took over low-level target control. The results might differ compared to plain JTAG/SWD
Info : clock speed 2000 kHz
Info : STLINK V2J25M14 (API v2) VID:PID 0483:374B
Info : Target voltage: 2.849302
Info : stm32f4x.cpu: Cortex-M4 r0p1 processor detected
Info : stm32f4x.cpu: target has 6 breakpoints, 4 watchpoints
Info : starting gdb server for stm32f4x.cpu on 0
Info : Listening on port 36371 for gdb connections
    TargetName         Type       Endian TapName            State       
--  ------------------ ---------- ------ ------------------ ------------
 0* stm32f4x.cpu       hla_target little stm32f4x.cpu       halted

Info : Unable to match requested speed 2000 kHz, using 1800 kHz
Info : Unable to match requested speed 2000 kHz, using 1800 kHz
target halted due to debug-request, current mode: Thread 
xPSR: 0x01000000 pc: 0x08000958 msp: 0x20000200
Info : device id = 0x20016419
Info : flash size = 2048 kbytes
Info : Dual Bank 2048 kiB STM32F42x/43x/469/479 found
auto erase enabled
wrote 16384 bytes from file /work/riot/RIOT/tests/touch_dev/bin/stm32f429i-disc1/tests_touch_dev.elf in 0.682543s (23.442 KiB/s)

verified 14652 bytes in 0.150211s (95.257 KiB/s)

Info : Unable to match requested speed 2000 kHz, using 1800 kHz
Info : Unable to match requested speed 2000 kHz, using 1800 kHz
shutdown command invoked
Done flashing
/work/riot/RIOT/dist/tools/pyterm/pyterm -p "/dev/ttyACM0" -b "115200"  
Twisted not available, please install it if you want to use pyterm's JSON capabilities
2022-01-13 13:09:38,915 # Connect to serial port /dev/ttyACM0
Welcome to pyterm!
Type '/exit' to exit.
2022-01-13 13:09:41,003 # Event: pressed!
2022-01-13 13:09:41,004 # X: 105, Y:158
2022-01-13 13:09:41,014 # X: 105, Y:158
2022-01-13 13:09:41,026 # X: 106, Y:157
2022-01-13 13:09:41,038 # X: 107, Y:158
2022-01-13 13:09:41,056 # X: 107, Y:158
2022-01-13 13:09:41,068 # X: 107, Y:158
2022-01-13 13:09:41,080 # X: 108, Y:158
2022-01-13 13:09:41,092 # X: 108, Y:158
2022-01-13 13:09:41,104 # X: 108, Y:158
2022-01-13 13:09:41,116 # Event: released!
2022-01-13 13:09:41,687 # Event: pressed!
2022-01-13 13:09:41,687 # X: 196, Y:141
2022-01-13 13:09:41,699 # X: 198, Y:143
2022-01-13 13:09:41,710 # X: 201, Y:142
2022-01-13 13:09:41,728 # X: 198, Y:141
2022-01-13 13:09:41,740 # X: 199, Y:142
2022-01-13 13:09:41,758 # X: 198, Y:142
2022-01-13 13:09:41,764 # X: 199, Y:142
2022-01-13 13:09:41,776 # X: 199, Y:142
2022-01-13 13:09:41,789 # Event: released!
2022-01-13 13:09:42,299 # Event: pressed!
2022-01-13 13:09:42,304 # X: 113, Y:76
2022-01-13 13:09:42,316 # Event: X: 111, Y:79
2022-01-13 13:09:42,328 # X: 111, Y:76
2022-01-13 13:09:42,341 # X: 111, Y:78
2022-01-13 13:09:42,352 # X: 111, Y:78
2022-01-13 13:09:42,370 # X: 112, Y:79
2022-01-13 13:09:42,382 # X: 112, Y:79
2022-01-13 13:09:42,394 # X: 112, Y:79
2022-01-13 13:09:42,406 # X: 112, Y:79
2022-01-13 13:09:42,418 # X: 112, Y:80
2022-01-13 13:09:42,431 # X: 112, Y:80
2022-01-13 13:09:42,442 # X: 112, Y:80
2022-01-13 13:09:42,460 # X: 113, Y:80
2022-01-13 13:09:42,472 # X: 112, Y:80
2022-01-13 13:09:42,490 # Event: released!
2022-01-13 13:09:42,724 # Event: pressed!
2022-01-13 13:09:42,725 # X: 91, Y:143
2022-01-13 13:09:42,736 # X: 91, Y:145
2022-01-13 13:09:42,748 # X: 90, Y:146
2022-01-13 13:09:42,760 # X: 90, Y:147
2022-01-13 13:09:42,772 # X: 91, Y:149
2022-01-13 13:09:42,790 # X: 90, Y:149
2022-01-13 13:09:42,803 # X: 90, Y:150
2022-01-13 13:09:42,814 # Event: released!
2022-01-13 13:09:43,650 # Exiting Pyterm
```

</details>

Reply:
Arf, I shouldn't have applied you proposal so quickly... Now I remember that the use of `choice` fails on Murdock and I don't know why.

Reply:
I reverted the change to the initial version of this PR otherwise it doesn't pass on Murdock.

Reply:
> Arf, I shouldn't have applied you proposal so quickly... Now I remember that the use of `choice` fails on Murdock and I don't know why.

Can you show the error message that triggered this?

Reply:
Is this issue showing up only on the CI? I quickly tested the suggestion locally building for the `stm32f429i-disc1` and seems to pass the hash check.

Reply:
The problem was with AVR based boards, at least.

Reply:
I'm trying to understand why the suggested change above should be used and why it's better. I'm also trying to understand why the version in master is not correct.

- Regarding the version in master: as I understand it, a choice element cannot be selected by another symbol and now that we introduce `HAVE_STMPE811_I2C`/`HAVE_STMPE811_SPI`, using choice is not possible (unless I'm missing something).

- Regarding the proposal above: the logic seems reversed to me and imply that the user would have to select both `MODULE_STMPE811` and (`MODULE_STMPE811_I2C` or `MODULE_STMPE811_SPI`). To me that doesn't fit.

Reply:
> * as I understand it, a choice element cannot be selected by another symbol

Yes, that's correct.

> using choice is not possible (unless I'm missing something).

The usage of `choice` is required due to the mutual exclusivity of the modules. The constraint is that `choice` always takes a value, unless it has the `optional` attribute or its dependencies are not met.

Using `optional` would not be an option (actually I think it was the initial approach for multi-variant drivers) because we still cannot 'activate' the choice with a `select` or so, then things like `SAUL_DEFAULT` or other options could not automatically enable modules. So we went with the approach of actually enabling the driver variant choice, and that's done with the module symbol directly.

> * Regarding the proposal above: the logic seems reversed to me and imply that the user would have to select both `MODULE_STMPE811` and (`MODULE_STMPE811_I2C` or `MODULE_STMPE811_SPI`). To me that doesn't fit.

The variant to use is just one more configuration of the module. There is always a default choice. And if the board already has it then no further configuration is needed.

Reply:
> I'm also trying to understand why the version in master is not correct.

We are still trying to sort out the design patterns. This is just the latest way that allows us to navigate around some corner-cases.

When I suggested my change I thought it should be mutually exclusive, only one interface would be used at a single time to connect to the device, which I guess is not correct.

Reply:
> When I suggested my change I thought it should be mutually exclusive, only one interface would be used at a single time to connect to the device, which I guess is not correct.

The current driver code is mutually exclusive, it does not handle having both, this is actually the case for all spi/i2c combos in master. There is no way to select the bus at runtime. So code would need to be changed for both to be possible.

Reply:
> > When I suggested my change I thought it should be mutually exclusive, only one interface would be used at a single time to connect to the device, which I guess is not correct.
> 
> The current driver code is mutually exclusive, it does not handle having both, this is actually the case for all spi/i2c combos in master. There is no way to select the bus at runtime. So code would need to be changed for both to be possible.

If this is the case then the bus variants should be a choice and @MrKevinWeiss proposal would be the way to go, right?

Reply:
> @MrKevinWeiss proposal would be the way to go, right?

Except that it doesn't pass the CI and seems reversed to me: as I understand it both `MODULE_STMPE811` and its variants must be selected. I would find it more natural to - from a user point of view - to only have to select a variant and have `MODULE_STMPE811` to be selected automatically.